core/lib/multivm/src/dump.rs

```diff
@@ -0,0 +1,379 @@
+use std::{collections::HashMap, fmt, sync::Arc};
+
+use serde::{Deserialize, Serialize};
+use zksync_types::{
+    block::L2BlockExecutionData, web3, L1BatchNumber, L2BlockNumber, Transaction, H256, U256,
+};
+use zksync_utils::u256_to_h256;
+use zksync_vm_interface::storage::WriteStorage;
+
+use crate::{
+    interface::{
+        storage::{InMemoryStorage, ReadStorage, StoragePtr, StorageView},
+        BootloaderMemory, BytecodeCompressionError, CompressedBytecodeInfo, CurrentExecutionState,
+        FinishedL1Batch, L1BatchEnv, L2BlockEnv, SystemEnv, VmExecutionMode,
+        VmExecutionResultAndLogs, VmFactory, VmInterface, VmInterfaceHistoryEnabled,
+        VmMemoryMetrics,
+    },
+    vm_fast, vm_latest, HistoryMode,
+};
+
+/// VM that tracks decommitment of bytecodes during execution. This is required to create a [`VmDump`].
+pub trait VmTrackingContracts: VmInterface {
+    /// Returns hashes of all decommitted bytecodes.
+    fn used_contract_hashes(&self) -> Vec<U256>;
+}
+
+impl<S: WriteStorage, H: HistoryMode> VmTrackingContracts for vm_latest::Vm<S, H> {
+    fn used_contract_hashes(&self) -> Vec<U256> {
+        self.get_used_contracts()
+    }
+}
+
+impl<S: ReadStorage> VmTrackingContracts for vm_fast::Vm<S> {
+    fn used_contract_hashes(&self) -> Vec<U256> {
+        self.decommitted_hashes().collect()
+    }
+}
+
+// FIXME: use storage snapshots (https://github.com/matter-labs/zksync-era/pull/2724)
```

Comment:
IDK whether it makes sense to wait for https://github.com/matter-labs/zksync-era/pull/2724 to get merged. Transition to using storage snapshots from there looks relatively straightforward; perhaps, the most non-trivial part is serialization logic.